rust/src/applayerpgsql/parser.rs

```diff
@@ -0,0 +1,155 @@
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+ use std::{fmt, self, u16, u8};
+ use nom::{number::complete::{be_u16, be_u32}};
+ 
+ // Dummy protocol major version, used for SSL, GSSAPI or Cancellation requests
+ pub const PGSQL_DUMMY_PROTO_MAJOR: u16 = 1234;
+ 
+ // Dummy protocol minor version, used for SSL encryption requests
+ pub const PGSQL_DUMMY_PROTO_MINOR_SSL: u16 = 5679;
+ 
+ #[derive(Debug, PartialEq)]
+ pub enum StartupMessageType {
+     StartupMessage = 0,
+     SslRequest = 1,
+     GssApiRequest = 2,
+ }
+ 
+impl fmt::Display for StartupMessageType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            StartupMessageType::StartupMessage => f.write_str("Startup Message"),
+            StartupMessageType::SslRequest => f.write_str("SSL Request"),
+            StartupMessageType::GssApiRequest => f.write_str("GSSAPI Request"),
+        }  
+    } 
+}
+
+ #[derive(Debug, PartialEq)]
+ pub struct PgsqlGenericStartupMessage{
+     message_type: StartupMessageType,
+     length: u32,
+     proto_major: u16,
+     proto_minor: u16,
+ }
+ 
+ named!(pub parse_ssl_request<PgsqlGenericStartupMessage>,
+     do_parse!(
+         len: verify!(be_u32, |&x| x == 8 )
+         >> proto_major: verify!(be_u16, |&x| x == PGSQL_DUMMY_PROTO_MAJOR )
+         >> proto_minor: verify!(be_u16, |&x| x == PGSQL_DUMMY_PROTO_MINOR_SSL )
+         >> (PgsqlGenericStartupMessage {
+             message_type: StartupMessageType::SslRequest,
+             length: len,
+             proto_major: proto_major,
+             proto_minor: proto_minor,
+         })
+ ));
+ 
+ pub const PGSQL_FLAG_SSL_NEGATIVE_RESPONSE: u8 = 'N' as u8;
+ pub const PGSQL_FLAG_SSL_POSITIVE_RESPONSE: u8 = 'S' as u8;
+ 
+ #[derive(Debug, PartialEq)]
+ pub struct PgsqlSslResponse {
+     message_type: u8,
+ }
+
+named!(pub parse_ssl_response<PgsqlSslResponse>,
+    do_parse!(
+        identifier: complete!(alt!(char!('N') | char!('S')))
+        >> (PgsqlSslResponse{
+            message_type: identifier as u8,
+        })
+    ));
+
+fn parse_len(input: &str) -> Result<u32, std::num::ParseIntError> {
+    input.parse::<u32>()
+}
+
+named!(pub parse_message<String>,
+       do_parse!(
+           len:  map_res!(
+                 map_res!(take_until!(":"), std::str::from_utf8), parse_len) >>
```

Comment:
I'm not understanding what is happening here. It parses the 4 byte length field, but then also takes something until a `:` is found? Shouldn't we just to something like `take!(len)`?

When testing with `postgresql-13-simple-select.pcap` I see this return `incomplete` all the time.

Reply:
Sorry, that's my fault, I haven't worked on this function, but now I can see that I certainly should, before sharing. (This also helps me understand when I can expect to see things "changing" if I run suricata after changing things here, and when not much will happen...) I will do that, and submit a new PR.